### PR TITLE
Reviewer leaderboard

### LINKED ISSUE #1250: Visible for users in the role Staff, create a "Reviewer leaderboard" over AC reviewers
Columns are:
* Reviewer
* Total
* 90 days
* This year
* Last year
* Most recent (date, links to most recent review if possible)

Allow sorting by those and the following filters:
* Funds
* Rounds
* Reviewers

Related to https://github.com/OpenTechFund/opentech.fund/issues/1249

### PR DESCRIPTION
**Pull request by ghickman**
This adds the Reviewer Leaderboard, a table of reviewers with agregated stats of their reviews the ability to filter on Reviewer, Fund, & Round.

Fixes #1250

### CODE REVIEW

**Comment by ghickman** [comment]
This doesn't currently have a link anywhere on the site, where should we link to it from?

File: opentech/apply/funds/views.py [hunk 1]
```diff
@@ -933,3 +938,33 @@ def handle_no_permission(self):
...
+        ninety_days_ago = timezone.now() - timedelta(days=90)
+        this_year = timezone.now().year
+        last_year = timezone.now().year - 1
+        latest_reviews = Review.objects.filter(author__reviewer_id=OuterRef('pk')).order_by('-created_at')
+        return super().get_table_data().filter(submissions_reviewer__isnull=False).annotate(
+            total=Count('assignedreviewers__review'),
+            ninety_days=Count('assignedreviewers__review', filter=Q(assignedreviewers__review__created_at__date__gte=ninety_days_ago)),
+            this_year=Count('assignedreviewers__review', filter=Q(assignedreviewers__review__created_at__year=this_year)),
+            last_year=Count('assignedreviewers__review', filter=Q(assignedreviewers__review__created_at__year=last_year)),
+            most_recent=Subquery(latest_reviews.values('id')[:1])
```
Comment (frjo) [inline opentech/apply/funds/views.py]: This query bogs down the database when run on live data. It needs to work for many thousands submissions/reviews.

**Comment by saoun** [comment]
@todd-dembrey can this be merged? anything holding it up?

**Comment by frjo** [comment]
@saoun The solution is to slow as it is now. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rudietuesdays** [comment]
@frjo is this live anywhere that we can look at? We're not able to get the branch running locally since it's so far behind master at this point. Happy to implement and advance the changes with @harumhelmy , but curious if there's somewhere we can look at it first before updating the code!

**Comment by frjo** [comment]
@rudietuesdays I have rebased this branch from master now so it's up to date.

File: hypha/apply/funds/views.py [hunk 2]
```diff
@@ -1157,3 +1162,33 @@ def get_submission_values(self):
...
+        if not (is_staff or is_reviewer):
+            raise PermissionDenied
+
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_table_data(self):
+        ninety_days_ago = timezone.now() - timedelta(days=90)
+        this_year = timezone.now().year
+        last_year = timezone.now().year - 1
+        # latest_reviews = Review.objects.filter(author__reviewer_id=OuterRef('pk')).order_by('-created_at')
```
Comment (frjo) [inline hypha/apply/funds/views.py:1187]: The comment out parts make this far too slow with live data. Better way to do this?

Could we base the query on the Review model instead and calculate review per user from that data?

File: hypha/apply/review/models.py [hunk 3]
```diff
@@ -101,6 +101,9 @@ def by_reviewers(self):
     def by_partners(self):
         return self.submitted()._by_group(PARTNER_GROUP_NAME)
 
+    def by_user(self, user):
+        return self.submitted().filter(author__reviewer=user).order_by('-created_at')
+
```
Comment (frjo) [inline hypha/apply/review/models.py:106]: Not sure this is needed by left it in after som experimentations.

File: hypha/apply/funds/views.py (hunk 2 above)
Comment (Parbhat) [reply hypha/apply/funds/views.py:1187]: Looks like the cause of slowness is in the `render_most_recent ` method. Adding a comment there.

File: hypha/apply/funds/tables.py [hunk 4]
```diff
@@ -358,3 +359,79 @@ class RoundsFilter(filters.FilterSet):
...
+                                .order_by('-created_at')
+                                .first())
+
+        if review is None:
+            return None
+
+        return format_html(
+            '<a href="{}">{}</a>',
+            review.revision.submission.get_absolute_url(),
+            review.revision.submission.title,
```
Comment (Parbhat) [inline hypha/apply/funds/tables.py:436]: This will make a DB query for every table row.

File: hypha/apply/funds/tables.py [hunk 5]
```diff
@@ -358,3 +359,79 @@ class RoundsFilter(filters.FilterSet):
...
+            'last_year',
+            'most_recent',
+        ]
+        model = User
+        order_by = ('-total',)
+
+    def render_most_recent(self, record):
+        review = (Review.objects.filter(author__reviewer=record)
+                                .order_by('-created_at')
+                                .first())
```
Comment (Parbhat) [inline hypha/apply/funds/tables.py:428]:
```suggestion
        review = (Review.objects.filter(author__reviewer=record)
                                .order_by('-created_at')
                                .select_related('revision__submission')).first()
```

File: hypha/apply/funds/views.py [hunk 6]
```diff
@@ -1157,3 +1162,33 @@ def get_submission_values(self):
...
+    filterset_class = LeaderboardFilter
+    table_class = LeaderboardTable
+    table_pagination = False
+    template_name = 'funds/review_leaderboard.html'
+
+    def dispatch(self, request, *args, **kwargs):
+        is_staff = request.user.is_apply_staff
+        is_reviewer = request.user.is_reviewer
+
+        if not (is_staff or is_reviewer):
```
Comment (Parbhat) [inline hypha/apply/funds/views.py:1178]: We can use UserPassesTestMixin as used in other [areas](https://github.com/OpenTechFund/hypha/blob/ceb74015590ae24d1412478505faeda6951c080d/hypha/apply/review/views.py#L61) of the codebase.

Comment (frjo) [reply hypha/apply/funds/views.py]: Fixed, thanks.

**Pull request merged**